pkg/impersonation/impersonation.go

```diff
@@ -182,10 +183,44 @@ func (i *Impersonator) createRole() (*rbacv1.ClusterRole, error) {
 			// in case cache isn't synced yet, use raw client
 			return i.clusterContext.RBAC.ClusterRoles("").Get(name, metav1.GetOptions{})
 		}
+		return r, nil
+	}
+	if !reflect.DeepEqual(rules, r.Rules) {
+		return i.updateRoleRules(r, rules[1], rules[2], rules[3])
 	}
 	return r, err
 }
 
+func (i *Impersonator) updateRoleRules(role *rbacv1.ClusterRole, groupRule, extrasPrincipalIDRule, extrasUsernameRule rbacv1.PolicyRule) (*rbacv1.ClusterRole, error) {
+	for i := range role.Rules {
+		if role.Rules[i].Resources[0] == "groups" {
+			for _, rn := range groupRule.ResourceNames {
+				role.Rules[i].ResourceNames = addToListIfNotPresent(rn, role.Rules[i].ResourceNames)
+			}
+		}
+		if role.Rules[i].Resources[0] == "userextras/principalid" {
+			for _, rn := range extrasPrincipalIDRule.ResourceNames {
+				role.Rules[i].ResourceNames = addToListIfNotPresent(rn, role.Rules[i].ResourceNames)
+			}
+		}
+		if role.Rules[i].Resources[0] == "userextras/username" {
+			for _, rn := range extrasUsernameRule.ResourceNames {
+				role.Rules[i].ResourceNames = addToListIfNotPresent(rn, role.Rules[i].ResourceNames)
+			}
+		}
+	}
+	return i.clusterContext.RBAC.ClusterRoles("").Update(role)
```

Comment:
Since this is additive and not making it equal, will that result in a situation in which each proxy API call you call update?  Basically first call `rules(A,B) != rules(C)` so you add C, then next call `rules(A, B, C) != rules(C)` so you don't modify the rule but still do an Update call?

Reply:
fixed